Initial implementation of the SRE Challenge (#60)
gravitational/careers: challenges/cloud/sre.md

[engineering levels document](https://raw.githubusercontent.com/gravitational/careers/main/levels.pdf)
for more details.

* Your submission will need to meet the requirements of the level you are applying for.
* Split the submission into 2-3 pull requests for us to review. We will review
  every pull request and provide our feedback.
* We are going to inspect the repository, test it, and get back to you.

## Server

An incomplete implementation of a Go server is located [here](sre-server.go).
You will extend this implementation by adding features described below depending
on the level you are applying to. Start by creating a new Github repository. Then
let the interview panel know the repository's location by pasting a link in your
interview Slack channel. Invite interview panel participants as contributors to
the new repository if you prefer to keep your submission private.

## Design Doc

Start with a brief design document that covers the edge cases and design approach. At
Teleport, we prefer Markdown for
[our designs](https://github.com/gravitational/teleport/blob/master/rfd/0000-rfds.md).

> Be sure to cover the following in your design:

Reviewer: Reading this top to bottom, and by the time I get here I still have very little context as to what the thing I'm designing is supposed to do. The only information I've been given so far is the initial sentence that mentions API and Kubernetes.

Would it make sense to add a few more sentences about the project here so the reader has this context in mind when reading about what needs to go in the design doc?

Reviewer 2: I see the issue that Zac is talking about. I'll add an intro section that summarizes the task.

> * Developer workflow

Reviewer: What does this mean? Can we be more specific?

Reviewer 3: Would adding a few bullet points like these be sufficient? I agree it's a bit vague:
  * Ease of contributing to the project from a fresh clone
  * Ease of building, running and testing the server
  * Ease of avoiding common mistakes

Author: Added the first two as subpoints, thank you!

* Level 3+: Build, Release
* Level 4+: State management and Delivery
* Level 5+: Reconciliation, Conflicts, and Automation

Open a pull request with your design and share a link with the reviewers via Slack.

> After the document is approved by at least 2 reviews, move on to the implementation.

Reviewer: ```suggestion
After the document is approved by at least 2 reviewers, move on to the implementation.
```

A few notes about the design document:

* We expect the design document to be completed roughly within the first week.
  This is to ensure you have enough time to work on the implementation.

> * Avoid writing an overly detailed design document. Two to three pages is
>   sufficient.

Reviewer 4: ```suggestion
* Avoid writing an overly detailed design document. 500-1500 words is
  sufficient.
```

Markdown doesn't neatly map to the pages, making 2-3 pages rather ambiguous.  I'm spitballing words (~500 per page) as an alternative length metric, but I'm open to other ideas.

Reviewer: Great suggestion Walt.

* Avoid sending us draft design documents, it is difficult to evaluate what
  parts are draft and which parts are complete. Instead we encourage asking
  questions in Slack and sharing a design document that is ready to be
  reviewed.

## Server & Automation

Once the design document is approved, begin working on adding features to the

> server implementation that includes automation. A key feature of this challenge 

Reviewer: Again, "add features that includes automation" doesn't mean much to me. 

So far, we've asked the candidate to design something and then to start adding features to the something but we have given almost no information as to what the something is.

is to produce a self-contained GitHub project that automates away as much as 
possible. This should include code quality, testing, build, and deployment.

Expect the interview panel to clone the repository and execute one or more 
Make targets that build and test a working solution. 

Add a couple of high quality tests that cover happy and unhappy scenarios. Keep
external dependencies as low or clear as possible. Expect the reviewers to
checkout the repository and execute the tests themselves.

Do not try to achieve full test coverage. This will take too long. Write enough
to exercise the different key components and show they are working as
intended, as well as to demonstrate your approach to automation.

## Deployment

Levels 3+ include a deployment step which should deploy the solution to a local

> Kubernetes cluster. The choice of which Kubernetes cluster is up to you, but 

Reviewer 4: > The choice of which Kubernetes cluster is up to you

This appears to conflict with later guidance to use AWS EKS.  If the intention is to have two deploy targets (e.g. minikube/kind _and_ EKS, that could be clarified).

Reviewer 2: We are only asking that they include an EKS section in the design.

please choose one that has the ability to run on macOS and Linux.

## Key Dependencies

You are welcome to solve this challenge using tools you are most familiar with,
but, at a minimum, please include the following:

* Go
* make
* Docker

## Level 1

### Server

> * HTTP API to retrieve the replica count of the Kubernetes Deployment

Reviewer: I think I see what's been confusing me. The header for "requirements" is on line 24, but we don't actually start talking about requirements for the project until this point. Everything up to now was more about guidance for the interview process and not actual requirements for the thing they need to build.

Maybe it's as simple as changing that first heading to something else so.

Reviewer 5: 👍 I found myself scrolling back up as I thought I missed the requirements.

### Automation

> * Ability to build the server in a Docker container

Reviewer: Does this mean write a Dockerfile for the server, or are we talking in a more general sense of some automation that produces an image? 

If it's the former and we want to see a Dockerfile, let's be explicit.

Also, the wording of this bullet point is a little weird. "build the server in a Docker container" could mean "build a runnable docker image" or it could mean "spin up a docker container that will run the build and produce an artifact that you can run outside of docker."

Author: I agree this is worded confusingly. 

The expectation would be simply to write a Dockerfile to build an image for this server, an implicit expectation when reviewing it would be to use a multi-stage build.

I've updated to (hopefully) be more clearer.

* Ability to run the server

## Level 2

### Server

* HTTP API to retrieve the replica count of the Kubernetes Deployment
* HTTP API to set the replica count of the Kubernetes Deployment

### Automation

* Ability to build the server in a Docker container
* Ability to run the server
* Ability to execute integration tests against the local Kubernetes cluster

## Level 3

### Server

* HTTP API to retrieve the replica count of the Kubernetes Deployment
* HTTP API to set the replica count of a the Kubernetes Deployment
* HTTP API to get the list of available Deployments in the Kubernetes cluster
* HTTP health check verifying Kubernetes connectivity

### Automation

* Ability to build the server in a Docker container
* Ability to run the server
* Ability to execute integration tests against the local Kubernetes cluster

> * GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass

Reviewer 4: > all tests pass

Does this refer to the integration tests from earlier? Are there expected to be any other tests? It'd be worth clarifying to `the integration tests pass` since integration tests in CI is a much bigger lift than `go test ./...` unit tests.

Author: Yes, there is an expectation to write more tests:

> Add a couple of high quality tests that cover happy and unhappy scenarios. Keep
external dependencies as low or clear as possible. Expect the reviewers to
checkout the repository and execute the tests themselves.

This isn't called out as a line item for each of the levels. So I'll add it there.

There's no expectation for a lot of test coverage, (1 or 2 tests should be okay) but expectations on the quality would increase.

Reviewer 4: Thanks for the clarification. Can we make sure to call out integration testing as explicitly in scope or out of scope for github actions?  I'd recommend:

* unit testing & static analysis/linting is in scope for CI
* integration testing is out of scope for CI

This is mostly to keep the challenge lighter.

Reviewer 3: How about just adding:

* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
  * Unit testing and static analysis is in scope
  * Integration testing can be omitted/skipped

Reviewer 4: > How about just adding:
> 
>     * GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
>       
>       * Unit testing and static analysis is in scope
>       * Integration testing can be omitted/skipped

Sounds good to me!

Author: Added - thanks!

### Deployment

* Create a helm chart for the service

## Level 4

### Design

In addition to including details about the requirements below, the design
document should include a section describing how this solution could be deployed
to an AWS EKS cluster.

### Server

* HTTP API to retrieve the replica count of the Kubernetes Deployment
* HTTP API to set the replica count of a the Kubernetes Deployment
* HTTP API to get the list of available Deployments in the Kubernetes cluster
* HTTP health check verifying Kubernetes connectivity
* Extend the API to support a difference between current and desired state
* Extend the Server to support the ability to show the differences between current and desired state

> * Secure connections with mTLS

Reviewer: The connection between the server and the k8s apiserver, the connection between the HTTP API and the caller, or both?

### Automation

* Ability to build the server in a Docker container
* Ability to run the server
* Ability to execute integration tests against the local Kubernetes cluster
* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass

> * Produce production ready releases (binaries and docker image)

Reviewer: This is pretty subjective and can mean different things to difference candidates depending on their background. Is there any way we can make this more fair by listing the specific things we'd like to see?

Author: > This is pretty subjective and can mean different things to difference candidates depending on their background.

This is a feature, not a bug

> Is there any way we can make this more fair by listing the specific things we'd like to see?

We could. We could also field those questions in slack, or the candidate can make assumptions. In any case, how the candidate interprets what it means to be production ready is an interesting signal to me. I'd rather keep this ambiguous so I can see what candidates come up with in the given time constraints rather than be explicit

Reviewer: I'll leave this up to you all then, but I'm not sure I agree.

Production ready to someone who's coming from Google is going to mean something different to someone who is maybe newer in their career, has a non-traditional background, or hasn't had the opportunity to work in big tech.

At the end of the day, both people are capable of doing the same work, I want to hire them both, not just the one who had the privilege to work somewhere where "production ready" more closely aligned with what our views.

Reviewer 2: Yeah, I think we're going to field some questions about this. We'll leave it in for now and update once we get a better idea of how to communicate without ambiguity.

### Deployment

* Create a helm chart for the service
* Deploy releases of the API Server to a Kubernetes cluster (and possible dependencies)
* Upgrade releases of the API Server to a Kubernetes cluster (and possible dependencies)

> * Include a production Deployment of this service, including but not limited to: Deployment, ServiceAccount, Service

Reviewer: This would likely be done via the helm chart mentioned above, right?

If so, I recommend listing the minimal criteria required for the helm chart at Level 3 (maybe it's just a deployment, and then add whatever else it is we care about here at level 4).

Author: Yeah great point, thank you. I've condensed so it's much more clearer.

I don't _believe_ there were extra requirements from L3 -> L4 with respects to Helm (@gozer, correct me if I'm wrong) so both requirements are mirrored.

## Level 5

### Design

In addition to including details about the requirements below, the design
document should include a section describing how this solution could be deployed
to an AWS EKS cluster. Please include security considerations as well.

### Server

* HTTP API to retrieve the replica count of the Kubernetes Deployment
* HTTP API to set the replica count of a the Kubernetes Deployment
* HTTP API to get the list of available Deployments in the Kubernetes cluster
* HTTP health check verifying Kubernetes connectivity
* Extend the API to support a difference between current and desired state
* Extend the Server to support the ability to show the differences between current and desired state
* Secure connections with mTLS

> * Replace the HTTP API with gRPC

Reviewer: In our challenges for Core, switching from. HTTP to gRPC was (IIRC) a level 3 item.

For SRE, maybe this makes sense to show that software engineering skills are more important at higher levels, but I also question whether this is just extra busy work that doesn't really differentiate level 4 and 5 candidates in any dimension other than how much time they have to work on it.

Author: > For SRE, maybe this makes sense to show that software engineering skills are more important at higher levels

This was a core part of why this was proposed, but there are other signals that I thought were important:

- securing via mTLS is expected to be more challenging, and can be more brittle compared to a web server
- We don't provide much in the way of a skeleton for gRPC, so it needs to be implemented from the ground up, rather than adapting what is already there
- how the protobuf contract is proposed is an interesting signal to me, it's more nuanced than HTTP, especially making a protobuf schema future proof 

In a nutshell, I'd probably make the distinguish the difference between the two as: An L4 candidate is expected to augment and adapt an existing codebase, whereas an L5 is expected to innovate and challenge the status quo.

With that said, I'm not super firm on getting these signals from gRPC -- though I do think it mirrors our working environment quite nicely

Reviewer 3: > In a nutshell, I'd probably make the distinguish the difference between the two as: An L4 candidate is expected to augment and adapt an existing codebase, whereas an L5 is expected to innovate and challenge the status quo.

I think that's describing it very well, exactly.

* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
* Produce production ready releases (binaries and docker image)
* Deploy and manage a local Kubernetes Cluster
* Deploy and upgrade to the local Kubernetes Cluster with no service interruption
* GitHub Action driven workflows for Deployment

### Deployment

* Create a helm chart for the service
* Deploy releases of the API Server to a Kubernetes cluster (and possible dependencies)
* Upgrade releases of the API Server to a Kubernetes cluster (and possible dependencies)
* Include a production Deployment of this service, including but not limited to: Deployment, Role, RoleBinding, ServiceAccount, Service

# Guidance

## Interview process

The interview team will join the Slack channel. The team consists of the
engineers who will be working with you. Ask them about the engineering culture,
work and life balance, or anything else that you would like to learn about
Teleport.

Before writing the actual code, create a brief design document in markdown in Github
and share with the team.

This document should cover at least: design approach, trade-offs, scope,

> proposed API, resiliency, scalability and deployment considerations.

Reviewer: This is somewhat repetitive of the original section, and makes the reader look in multiple places, For example:

- we talk about starting with a design document on line 58 and then introduce the design doc again here
- we mention splitting the code into 2-3 PRs on line 31 but then we talk about splitting into PRs again on line 239

Author: I don't disagree that in parts this section doesn't introduce new information, but as long as we are consistent I don't think it's such an issue to reinforce and condense our expectations under the Guidance section, particularly if it's just a single line. 

With that said, I understand the value of brevity, so I've pared down this to focus on the process and less on expectations, PTAL

  provide feedback on the various implementation phases. We are a distributed
  team, so structured, asynchronous communication is critical to us.

## Questions
 
It is OK to ask the interview team questions. Some folks stay away from asking
questions to avoid appearing less experienced, so we provide examples of
questions to ask and questions we expect candidates to figure out on their own.

Here is a great question to ask:

> Is it OK to assume there will be only a single target kubernetes cluster for
  this service? I will add a note on how support for multiple clusters could be
  implemented, but it feels like an unnecessary complexity.

This is the question we expect candidates to figure out on their own:

> What version of Go should I use? What dependency manager should I use? What
  framework/tool should I use to automate testing and deployment ?

Unless specified in the requirements, pick the solution that works best for you.

# Tools

This task should be implemented in Go and should work on a 64-bit Linux or MacOS
machine.

It's safe to assume a working Docker environment will be available locally as well.

> Additional external dependencies are acceptable such as Terraform and minikube/k3d,
> but please ensure your Makefile is handles these dependencies.

Reviewer 4: ```suggestion
Additional external dependencies are acceptable such as Terraform and minikube/k3d,
but please ensure your Makefile has targets to install these dependencies.
```

Reviewer: Seems fine to just have the makefile spit out "terraform: command not found" as long as instructions for how to install it are linked, don't you think?

Make is a bit esoteric and writing makefiles to install tools that could easily be installed via brew (for example) doesn't seem the best use of a candidate's time.

Reviewer 4: That's fair.  This change was motivated more out of a desitre be clear when we say the `Makefile` should "handle" the dependencies.  Im happy with either an installation target or context aware failures as long as it is clear for the candidate what is expected.

# Timing

You can split coding over a couple of weekdays or weekends and find time to ask
questions and receive feedback.

> Once you join the Slack channel, you have between 1 to 2 weeks complete the
> challenge depending on the challenge you choose.

Reviewer 4: ```suggestion
Once you join the Slack channel, you have up to 2 weeks to complete the
challenge.
```

There doesn't appear to be a choice of challenges presently.

to exercise the different key components and show they are working as
intended, as well as to demonstrate your approach to automation.

## Deployment

Levels 3+ include a deployment step which should deploy the solution to a local
Kubernetes cluster. The choice of which Kubernetes cluster is up to you, but 
please choose one that has the ability to run on macOS and Linux.

## Key Dependencies

You are welcome to solve this challenge using tools you are most familiar with,
but, at a minimum, please include the following:

* Go
* make
* Docker

## Level 1

### Server

* HTTP API to retrieve the replica count of the Kubernetes Deployment
* One or two tests that cover happy and unhappy scenarios

### Automation

* Write a Dockerfile to build an image for the server
* Ability to run the server

## Level 2

### Server

* HTTP API to retrieve the replica count of the Kubernetes Deployment
* HTTP API to set the replica count of the Kubernetes Deployment
* One or two tests that cover happy and unhappy scenarios

### Automation

* Write a Dockerfile to build an image for the server
* Ability to run the server
* Ability to execute integration tests against the local Kubernetes cluster

> * One or two tests that cover happy and unhappy scenarios

Reviewer 2: Looks like a dup.

### Automation

* Write a Dockerfile to build an image for the server
* Ability to run the server
* Ability to execute integration tests against the local Kubernetes cluster
* GitHub Actions for every commit that verifies code quality, ensures the software builds, and all tests pass
  * Unit testing and static analysis is in scope
  * Integration testing can be omitted/skipped

### Deployment

* Create a helm chart for the service that includes, a minimum: a Deployment, ServiceAccount and Service

## Level 4

### Design

In addition to including details about the requirements below, the design
document should include a section describing how this solution could be deployed
to an AWS EKS cluster. Please note that you are not expected to implement any
deployment automation for an EKS target, or create any AWS resources.

### Server

* HTTP API to retrieve the replica count of the Kubernetes Deployment
* HTTP API to set the replica count of a the Kubernetes Deployment
* HTTP API to get the list of available Deployments in the Kubernetes cluster
* HTTP health check verifying Kubernetes connectivity
* Extend the API to support a difference between current and desired state

> * Extend the Server to support the ability to show the differences between current and desired state (implementation of state storage for the server is left as a design choice for the candidate)

Reviewer 5: What's the difference between both points?

Reviewer 2: Yeah, I had the same question in the original version. We've updated it since but I agree that there is still too much overlap.
